packages/iTwinUI-react/src/core/Table/actionHandlers/selectHandler.ts

```diff
@@ -56,6 +56,109 @@ export const onSelectHandler = <T extends Record<string, unknown>>(
   onSelect?.(selectedData, newState);
 };
 
+/**
+ * Handles selection when clicked on a row while shift is selected.
+ */
+export const onShiftSelectHandler = <T extends Record<string, unknown>>(
+  state: TableState<T>,
+  action: ActionType,
+  instance: TableInstance<T>,
+) => {
+  const resetSelection = () => {
+    instance.toggleAllRowsSelected(false);
+  };
+
+  const getGreaterId = (id: string) => {
+    const greaterId = id.split('.');
+    greaterId[greaterId.length - 1] = `${
+      parseInt(greaterId[greaterId.length - 1]) + 1
+    }`;
+    return greaterId.join('.');
+  };
+
+  const getParentId = (id: string) => {
+    const lastDotIndex = id.lastIndexOf('.');
+    if (lastDotIndex < 0) {
+      return null;
+    }
+    return id.substring(0, lastDotIndex);
+  };
+
+  /**
+   * @returns `true` if `subId` is indeed a sub-row of `id`, else `false`
+   * @example
+   * isSubRow('1.4', '1.4.2.3') // true
+   * isSubRow('1.3', '1.2.5')   // false
+   * isSubRow('2.4.5', '2.4.5') // true
+   */
+  const isSubRow = (id: string, subId: string) => {
+    const idSplit = id.split('.');
+    const subIdSplit = subId.split('.');
+
+    let i = 0;
+    while (i < idSplit.length) {
+      if (idSplit[i] !== subIdSplit[i]) {
+        return false;
+      }
+      i++;
+    }
+
+    return true;
+  };
+
+  /** Returns the next row as if it was just a non sub-rows list */
+  const getNextRow = (currentRow: Row<T>, endId: string) => {
+    let nextRow;
+    if (
+      (currentRow.subRows ?? []).length > 0 &&
+      isSubRow(currentRow.id, endId)
+    ) {
+      // child
+      nextRow = currentRow.subRows[0];
+    } else {
+      // next sibling
+      nextRow = instance.rowsById[getGreaterId(currentRow.id)];
+
+      // If the next sibling doesn't exist, go to the next child at a one upper depth
+      let parentId = getParentId(currentRow.id);
+      while (nextRow == null && parentId != null) {
+        nextRow = instance.rowsById[getGreaterId(parentId)];
+        parentId = getParentId(parentId);
+      }
+    }
+
+    return nextRow;
+  };
+
+  resetSelection();
+
+  // If no row selected before shift clicking, then assume first row is the last selectedRow (like Windows File Explorer)
+  let startId = state.lastSelectedRow ?? instance.rows[0].id;
+  let endId = action.id;
+
+  // Make sure startId is lesser than endId. Selection always goes from top to bottom
+  if (startId > endId) {
+    const temp = startId;
+    startId = endId;
+    endId = temp;
+  }
+
+  let currentRow = instance.rowsById[startId];
+
+  while (currentRow != null) {
+    if (currentRow.id === endId) {
+      currentRow.toggleRowSelected(true);
```

Comment:
Thanks, yes, being able to dispatch a list of selected rows in one call sounds way better. However, I'm a little confused on how to implement it. Because currently also we are doing something similar. On a shift click, we are dispatching a table action for shift click that only contains the clicked row (e.g. the end row) in the action.

###### Table.tsx
![image](https://user-images.githubusercontent.com/45748283/187690026-882a2e31-ad51-45e0-b1ae-f14acb6c9897.png)

The start row is saved in the state. So the action handler (`onShiftSelectHandler`) can get the start and end rows and toggle them one by one.

###### selectHandler.ts
![image](https://user-images.githubusercontent.com/45748283/187689171-743e2463-c3a1-4f08-8b87-d827572f035b.png)

What part of this process can I make more efficient? Maybe the code in the handler (`onShiftSelectHandler`) to toggle multiple rows together? But I'm not sure how to do that.

Also, just to mention, the handler (`onShiftSelectHandler`) doesn't blindly toggle all the rows between the start and the end. If it knows that all the subrows are going to be toggled anyways, it just toggles the row itself and skips toggling the subrows.

Reply:
i think [`selectedFlatRows`](https://react-table-v7.tanstack.com/docs/api/useRowSelect#instance-properties:~:text=will%20be%20false-,selectedFlatRows,-%3A%20Array%3CRow%3E) might help

Reply:
i think [`selectedFlatRows`](https://react-table-v7.tanstack.com/docs/api/useRowSelect#instance-properties:~:text=will%20be%20false-,selectedFlatRows,-%3A%20Array%3CRow%3E) might help

Reply:
> i think [`selectedFlatRows`](https://react-table-v7.tanstack.com/docs/api/useRowSelect#instance-properties:~:text=will%20be%20false-,selectedFlatRows,-%3A%20Array%3CRow%3E) might help

Thanks for the idea, but I'm not sure how you meant I should use that. Because I think it stores all the previously selected rows/subrows. But I'm not sure how it can help avoid toggling all the row/subrows between the start and the end.

I thought that one benefit might be that when resetting the row selections (`resetSelection()`), we can only deselect rows that are not selected in the new selection? E.g. old selection = `1, 2, 3, 4, 5`; new selection = `1, 2, 3`; then only deselect 4 and 5. But I then thought that this might be more inefficient for extremely large tables (1000+ rows) than the idea you suggested of `instance.toggleAllRowsSelected(false);`.

Reply:
**[DONE]** Thanks for the ideas Vilius and Mayank. As mentioned here ([#774 comment](https://github.com/iTwin/iTwinUI-react/pull/774/files/da5ef1854d02894e3c653eaf158ee178b75bd210#r968990008)), saving our own `Record / {}` and setting the state at the end allowed `onSelect()` to be called only once per shift click, instead of once per toggle.